builtin/logical/database/dbplugin/client.go

```diff
@@ -18,13 +19,25 @@ type DatabasePluginClient struct {
 	Database
 }
 
+// This wraps the Close call and ensures we both close the database connection
+// and kill the plugin.
 func (dc *DatabasePluginClient) Close() error {
 	err := dc.Database.Close()
 	dc.client.Kill()
 
 	return err
 }
 
+// This wraps the Initialize call and ensures we close the plugin on error.
+func (dc *DatabasePluginClient) Initialize(ctx context.Context, config map[string]interface{}, verifyConnection bool) error {
+	err := dc.Database.Initialize(ctx, config, verifyConnection)
+	if err != nil {
+		dc.Close()
```

Comment:
The database object in here might be nil if `verify_connection` was set to false (thus never calling `Connection()` to instantiate the object). dc.Close() -> dc.Database.Close() would be called on a nil db if the Initialize error comes from something else (e.g. https://github.com/hashicorp/vault/blob/master/plugins/helper/database/connutil/sql.go#L58).

Reply:
Isn't that something that should be handled by the Close() implementation in the database though?

Reply:
In fact isn't it handled already? https://github.com/hashicorp/vault/blob/master/plugins/helper/database/connutil/sql.go#L132

Reply:
Yeah the plugin implementation handles that. https://github.com/hashicorp/vault/blob/master/plugins/helper/database/connutil/sql.go#L127

Reply:
Ah, missed that. It's all good here!